enarx-keep-sgx-shim/src/start.S

```diff
@@ -56,26 +117,112 @@ shp:                            # The bottom of the secure heap
     .globl enclave
     .type enclave, @function
 enclave:
-    lea     stk(%rip),              %rdi            # Get the stack address
-    xchg    %rsp,                   %rdi            # Swap stacks
-    push    %rdi                                    # Store the old stack
-    push    %rcx                                    # Store next non-enclave addr
-
-    cmp     $0,                     %rax            # If CSSA == 0 ...
-    jne     .event                                  # ... go to .event
-
-.entry:
-    ud2                                             # Synthetic bad instruction
-    jmp     .eexit                                  # Do EEXIT
-
-.event:
-    mov     %rax,                   %rdi            # %rdi = CSSA
-    shl     $12,                    %rdi            # %rdi = CSSA * 4096
-    add     %rbx,                   %rdi            # %rdi = aex[CSSA - 1]
-    call    event                                   # Call our event handler
-
-.eexit:
-    pop     %rbx                                    # Restore next non-enclave addr
-    pop     %rsp                                    # Restore old stack
-    mov     $4,                     %rax            # EEXIT
-    enclu                                           # ENCLU[EEXIT]
+    xchg    %rbx,                   %rcx            # Swap TCS and next instruction.
+    cmp     $0,                     %rax            # If CSSA > 0...
+    jne     .Levent                                 # ... restore stack from AEX[CSSA-1].
+
+# CSSA == 0
+    rdrand  %r11                                    # Get a random number
+    and     $0xff0,                 %r11            # Mask to an aligned random offset
+    lea     stk(%rip),              %r10            # Get the stack pointer
+    sub     %r11,                   %r10            # Subtract random offset
+
+    xor     %r11,                   %r11            # %r11 = NULL
+    xor     %rax,                   %rax            # %rax = NULL
+    jmp     .Lstack
+
+# CSSA != 0
+.Levent:
+    dec     %rax                                    # %rax = CSSA - 1
+    shl     $12,                    %rax            # %rax = (CSSA - 1) * 4096
```

Comment:
hm, not entirely sure, but I think you're silently assuming here that SSA frames are 1page (4KiB) in size(?) AFAIK this is _not_ guaranteed in the SGX architecture and you'd have to retrieve that from the SECS.SSAFRAMESIZE attribute:

> SSAFRAMESIZE 16 4 Size of one SSA frame in pages, including XSAVE, pad, GPR, and MISC (if
CPUID.(EAX=12H, ECX=0):.EBX != 0).

In practice, it might be that the way you build enclaves in Enarx they always have an SSA frame size of 1page (for now), but then again I think you should better explicitly document this assumption and encode it in a constant?

Reply:
I'll address this.

Currently, the Enarx SGX libraries force the frame size to one 4K page. There is also, unfortunately, no access to the SECS from inside the enclave. Once we get EDMM support, all of this code will become dynamic. So the problem gets "solved" implicitly in the long run. But for now I can at least enshrine this in a constant.